### PR TITLE
Support full unicode in parser

### PR DESCRIPTION
**Pull request by dondonz**
This PR implements the RFC to support full Unicode in the parser.

## Key spec changes
+ GraphQL now supports a wider range of Unicode characters. `SourceCharacter` was expanded to include any Unicode code point that is neither a leading nor trailing surrogate. Previously only up to U+FFFF included
+ Spec now includes guidance on Unicode surrogate pairs
+ (minor) GraphQL now allows certain control characters

## Key changes in this PR
This PR has two halves:
1. `UnicodeUtil`, used by `StringValueParsing`. This is to handle braced escapes and escaped surrogate pairs
2. ANTLR grammar changes, which expand the definition of a `SourceCharacter` to mean any Unicode code point that is neither a leading nor trailing surrogate

### References
**RFC GitHub issue:** https://github.com/graphql/graphql-spec/issues/687
**RFC spec text:** https://github.com/graphql/graphql-spec/pull/849
**RFC JS implementation:** https://github.com/graphql/graphql-js/pull/3117
**Previous PR:** https://github.com/graphql-java/graphql-java/pull/2335

---
_Want a Unicode fun fact?_ Groovy fails to compile if there are any Unicode code points that are not exactly four hex digits. You'll even encounter this compilation problem in COMMENTS.

For example: this comment containing RFC text will cause a compilation error
```
For example the input `"\uD83D\uDCA9"` is a valid {StringValue} which represents the same Unicode text as `"\u{1F4A9}"`.
```

The fix is to add an extra backslash
```
... which represents the same Unicode text as `"\\u{1F4A9}"`.
```

### CODE REVIEW

File: src/main/java/graphql/parser/UnicodeUtil.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                } catch (IOException e) {
+                    return Assert.assertShouldNeverHappen();
+                }
+                return endIndexExclusive;
+            } else {
+                throw new RuntimeException("invalid unicode code point");
+            }
+        }
+//        Assert.assertShouldNeverHappen();
+        // TODO error checking of invalid values
```
Comment (bbakerman) [inline src/main/java/graphql/parser/UnicodeUtil.java:53]: Fix this up

File: src/main/java/graphql/parser/UnicodeUtil.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            if (isValidUnicodeCodePoint(hexValue)) {
+                char[] chars = Character.toChars(hexValue);
+                try {
+                    writer.write(chars);
+                } catch (IOException e) {
+                    return Assert.assertShouldNeverHappen();
+                }
+                return endIndexExclusive;
+            } else {
+                throw new RuntimeException("invalid unicode code point");
```
Comment (bbakerman) [inline src/main/java/graphql/parser/UnicodeUtil.java:49]: Put the offening value in the exception - make this a GraphqlException

Comment (bbakerman) [reply src/main/java/graphql/parser/UnicodeUtil.java:49]: Or will this get re-translated by ANTLR into something else

We want  a graphql.parser.InvalidSyntaxException at the end

File: src/test/groovy/graphql/parser/UnicodeUtilParserTest.groovy [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+              {
+              foo(arg: "\\uD83D")
+               }
+        '''
+
+        when:
+        Document document = Parser.parse(input)
+
+        then:
+        // TODO: Discuss whether to raise exception
```
Comment (bbakerman) [inline src/test/groovy/graphql/parser/UnicodeUtilParserTest.groovy:141]: notThrown is how you do this in Spock

File: src/test/groovy/graphql/parser/UnicodeUtilParserTest.groovy [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+              {
+              foo(arg: "\\uD83D\\uDBFF")
+               }
+        '''
+
+        when:
+        Document document = Parser.parse(input)
+
+        then:
+        // TODO: Raise exception
```
Comment (bbakerman) [inline src/test/groovy/graphql/parser/UnicodeUtilParserTest.groovy:110]: notThrown??

File: src/main/java/graphql/parser/UnicodeUtil.java (hunk 2 above)
Comment (dondonz) [reply src/main/java/graphql/parser/UnicodeUtil.java:49]: There are two sides of this Unicode RFC:
1. Unescaped unicode characters like 🍺
2. Escaped unicode characters such as `\u{1F37A}` (escaped version of 🍺)

Unescaped unicode characters are parsed by ANTLR and will raise the usual `InvalidSyntaxException`.

Escaped unicode characters are parsed separately by this `UnicodeUtil`, which is called by `parseSingleQuotedString()`. As this is outside ANTLR, exceptions are not automatically re-translated - I'll add this.

**Review by bbakerman — APPROVED** [review]

File: src/main/antlr/GraphqlCommon.g4 [hunk 5]
```diff
@@ -120,12 +120,12 @@ fragment BlockStringCharacter:
 ExtendedSourceCharacter;
 
 fragment StringCharacter:
-([\u0009\u0020\u0021] | [\u0023-\u005b] | [\u005d-\u{10FFFF}]) |  // this is SoureCharacter without '"' and '\'
+([\u0009\u0020\u0021] | [\u0023-\u005b] | [\u005d-\u{10FFFF}]) |  // this is SourceCharacter without '"' and '\'
```
Comment (andimarek) [inline src/main/antlr/GraphqlCommon.g4:123]: this is not completely correct: we want to allow all code points except surrogates: this  still disallows certain control characters.

**Review by andimarek — APPROVED** [review]

**Review by bbakerman — APPROVED** [review]

**Pull request merged**